util/network-devp2p/src/math.rs

```diff
@@ -0,0 +1,97 @@
+// Copyright 2018 Coinbase, Inc.
+// This file is part of Parity.
+
+// Parity is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+/// Trait defines base 2 log operation on unsigned integer types, rounded down to the nearest
+/// integer.
+pub trait FastLog {
+	fn log2(v: Self) -> usize;
+}
+
+macro_rules! fast_log2_impl {
+	($t:ty, $log_log_limit:expr) => (
+		// Efficiently implemented using bit manipulations to binary search for the index of most
+		// significant bit set.
+		//
+		// See: http://graphics.stanford.edu/~seander/bithacks.html#IntegerLog
+		impl FastLog for $t {
+			#[inline]
+			fn log2(v: $t) -> usize {
+				if v == 0 {
+					panic!("FastLog::log2 cannot be called with 0");
+				}
+
+				let b: [u64; 6] = [0x2, 0xC, 0xF0, 0xFF00, 0xFFFF0000, 0xFFFFFFFF00000000];
+				let s: [usize; 6] = [1, 2, 4, 8, 16, 32];
```

Comment:
Maybe change `b` and `s` to static so that we don't need to allocate stack each time `log2` is called?